File: leetcode_49.cpp
//把anagrams-- 由同样字母组成的不同单词 放到同一个vector中
//核心是构造一个map， 以排序过的字符串做key

class Solution {
public:
    map<string, vector<string>> M;
    void fufill_M(vector<string>& strs){
        for(auto &str: strs){
            string str_back = str;
            sort(str_back.begin(), str_back.end());
            M[str_back].push_back(str);
        }
    }

    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        fufill_M(strs);
        vector<vector<string>> result;
        for(auto i=M.begin(); i!=M.end(); i++){
            result.push_back(i->second);
        }
        return result;
    }
        
};

File: leetcode_49.py
#list 是不可hash的， tuple可哈希， 具体是因为tuple不可变
import collections
class Solution(object):
    def groupAnagrams(self, strs):
        ans = collections.defaultdict(list)  # 创建一个每一项value都是一个list的dict
        for s in strs:
            ss = sorted(s)   #ss是一个按照字母顺序排列的list，每个元素是一个字母
            ans[tuple(ss)].append(s)  # tuple(str)能构建一个str的每一个字符为元素构成的tuple.
        return ans.values()


